### PR TITLE
Warning when editing multiple constraint sets, and ability to copy them

### PR DESCRIPTION
**Pull request by rpiaggio**
* When in Observations View: a warning is shown when attempting to edit a constraint set shared by multiple observations, with the option to copy them or dismiss the warning.
* When in Constraints View: a copy button is shown in the title bar.
* Copy is working.
* Loaders everywhere are now nicer SUI loaders.

This is PoC. We can change the mechanism, and or adapt to other tiles/views.

(Please feel free to restyle at leisure)

Kapture 2021-05-19 at 21 53 25

Kapture 2021-05-19 at 21 55 51

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -154,7 +154,7 @@ lazy val commonSettings = lucumaGlobalSettings ++ Seq(
   publishLocal := {},
   publishArtifact := false,
   Keys.`package` := file(""),
-  scalacOptions --= Seq("-Xfatal-warnings").filter(_ => insideCI.value)
+  scalacOptions --= Seq("-Xfatal-warnings").filterNot(_ => insideCI.value)
```
Comment (rpiaggio) [inline build.sbt:157]: This is an unrelated fix. Enjoy non-fatal warnings in dev.

File: common/src/main/scala/explore/components/Tile.scala [hunk 2]
```diff
@@ -29,7 +33,8 @@ final case class Tile(
   canMinimize:       Boolean = false,
   canMaximize:       Boolean = false,
   state:             TileSizeState = TileSizeState.Normal,
-  sizeStateCallback: TileSizeState ~=> Callback = Reusable.always(_ => Callback.empty)
+  sizeStateCallback: TileSizeState ~=> Callback = Reusable.always(_ => Callback.empty),
+  key:               js.UndefOr[Key] = js.undefined
```
Comment (rpiaggio) [inline common/src/main/scala/explore/components/Tile.scala:37]: Tiles can now keyed. This is used to force a remount when switching an observation but one of the tiles references the same object as the previous observation.

File: model/shared/src/main/scala/explore/model/ElevationRange.scala [hunk 3]
```diff
@@ -4,17 +4,39 @@
...
 import io.circe.Encoder
+import io.circe._
 import io.circe.generic.semiauto._
 import io.circe.refined._
 import lucuma.core.optics.SplitEpi
 import monocle.Prism
 import monocle.macros.GenPrism
 import monocle.macros.Lenses
 
+private sealed abstract class ElevationRangeType(val typeName: String) { self =>
```
Comment (rpiaggio) [inline model/shared/src/main/scala/explore/model/ElevationRange.scala:20]: `ConstraintSetModel` was removed in favor of the generated classes from the query. We still use these classes to model ElevationRange, so decoders were moved here.

**Review by cquiroz — APPROVED** [review]
Thanks, we should let science test this.  let's  test the concept with users and then we can tweak the UI

File: explore/src/main/scala/explore/Explore.scala [hunk 4]
```diff
@@ -50,7 +50,7 @@ import js.annotation._
 
 @JSExportTopLevel("Explore")
 object ExploreMain {
-  japgolly.scalajs.react.extra.ReusabilityOverlay.overrideGloballyInDev()
+  // japgolly.scalajs.react.extra.ReusabilityOverlay.overrideGloballyInDev()
```
[the remark this answers is not in the document]
Comment (rpiaggio) [reply explore/src/main/scala/explore/Explore.scala]: yup, aready fixed in later commit

File: explore/src/main/scala/explore/constraints/ConstraintsPanel.scala [hunk 5]
```diff
@@ -138,13 +156,106 @@ object ConstraintsPanel {
...
 
-      ReactFragment(
+      val obsCount        = props.constraintSet.get.observations.totalCount
+      val showWarningView = state.zoom(State.showMultiEditWarning.composePrism(some))
+
+      val makeCopy =
+        for {
+          _  <- state.zoom(State.copying).set(true)
+          _  <- showWarningView.set(false)
+          id <- IO(Random.nextInt(0xfff)).map(int =>
```
[the remark this answers is not in the document]
Comment (rpiaggio) [reply explore/src/main/scala/explore/constraints/ConstraintsPanel.scala:166]: yes

File: common/src/main/scala/explore/implicits.scala [hunk 6]
```diff
@@ -190,4 +193,18 @@ object implicits
...
+  // React implicits
+  implicit class HtmlAttrsOps(val a: HtmlAttrs) extends AnyVal {
+    // Generalize https://gist.github.com/pstoica/4323d3e6e37e8a23dd59
+    def onComponentBlur(handler: Callback): TagMod =
+      a.onBlur ==> { (e: ReactFocusEvent) =>
+        val currentTarget = e.currentTarget
+
+        handler
+          .when_(!currentTarget.contains(dom.document.activeElement))
+          .setTimeout(duration.Duration.Zero)
```
[the remark this answers is not in the document]
Comment (rpiaggio) [reply common/src/main/scala/explore/implicits.scala:206]: The zero timeout forces the callback (including checking the condition in `when_`) to be executed after the current event stack is processed. This guarantees that `activeElement` is updated.

**Comment by rpiaggio** [comment]
> Cool, should the key be based on the obs id? Maybe the id should reflect that but it would cause the user properties be per obsid which maybe overkill

Not sure I understand the suggestion. In the case of the observations view, the key is the obs id. But now on second thought, it should be a combination of the obs id and the id of whatever is displayed in the tile, otherwise we can have duplicate keys. I'll fix that.

**Pull request merged**